hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/clustering/run/strategy/SparkSingleFileSortExecutionStrategy.java

```diff
@@ -77,8 +78,11 @@ public HoodieData<WriteStatus> performClusteringWithRecordsAsRow(Dataset<Row> in
     // Since clustering will write to single file group using HoodieUnboundedCreateHandle, set max file size to a large value.
     newConfig.setValue(HoodieStorageConfig.PARQUET_MAX_FILE_SIZE, String.valueOf(Long.MAX_VALUE));
 
-    return HoodieDatasetBulkInsertHelper.bulkInsert(inputRecords, instantTime, getHoodieTable(), newConfig,
-        getRowPartitioner(strategyParams, schema), numOutputGroups, shouldPreserveHoodieMetadata);
+    BulkInsertPartitioner<Dataset<Row>> partitioner = getRowPartitioner(strategyParams, schema);
+    Dataset<Row> repartitionedRecords = partitioner.repartitionRecords(inputRecords, numOutputGroups);
```

Comment:
Doesn't this already happen inside `HoodieDatasetBulkInsertHelper.bulkInsert`?

Reply:
Yes, I changed `HoodieDatasetBulkInsertHelper.bulkInsert`, deleting repartition in that method. so adding repartition here and `SparkSortAndSizeExecutionStrategy`

``` scala
// deleted
    val repartitionedDataset = partitioner.repartitionRecords(dataset, parallelism)
    val arePartitionRecordsSorted = partitioner.arePartitionRecordsSorted
```

`BaseDatasetBulkCommitActionExecutor` would first call `HoodieDatasetBulkInsertHelper.prepareForBulkInsert`, and then call `HoodieDatasetBulkInsertHelper.bulkInsert` to perform write operation. And in `HoodieDatasetBulkInsertHelper.prepareForBulkInsert`, it would do the repartition work, so I delete the repartition in `HoodieDatasetBulkInsertHelper.bulkInsert`. And it looks more reasonable, `HoodieDatasetBulkInsertHelper.bulkInsert` should only perform the write operation, it shouldn't change the records layout.

Reply:
Got it. Thanks for the explanation.